govc: add support for supervisor services deploy

rikatz opened the pull request:
Signed-off-by: Ricardo Katz <rkatz@vmware.com>

## Description

Initial work on supporting vSphere Namespaces supervisor services.

Right now it's possible to:
* List
* Enable
* Disable
* Remove
* Create


TODO: Write VCSIM things

## Type of change

Please mark options that are relevant:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update
- [ ] Build related change

## How Has This Been Tested?

Tested with a real cluster. Still will check to write some unit tests

## Checklist:

- [ ] My code follows the CONTRIBUTION [guidelines](./../CONTRIBUTING.md) of
  this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

embano1 commented on govc/USAGE.md:
```diff
@@ -230,6 +230,12 @@ but appear via `govc $cmd -h`:
  - [namespace.cluster.enable](#namespaceclusterenable)
  - [namespace.cluster.ls](#namespaceclusterls)
  - [namespace.logs.download](#namespacelogsdownload)
+ - [namespace.supervisorsvc.activate](#namespacesupervisorsvcactivate)
```
question: `supervisorsvc` or more user-friendly just `supervisor`?

rikatz replied:
Hum, depends actually on the other APIs for a future:

I was trying to follow the API pattern for the service names (but without having something like `supervisorservice`, but per-api this endpoint is https://developer.vmware.com/apis/vsphere-automation/latest/vcenter/namespace_management/supervisor_services/

And then we probably will have in a future something like `supervisorsvc.cluster` to add/enable/disable per clusters (https://developer.vmware.com/apis/vsphere-automation/latest/vcenter/namespace_management/supervisor_services.cluster_supervisor_services/)

and finally the versions (`supervisorsvc.versions`).

So We may end up with a huge tree like:
- namespace.supervisorsvc.cluster.create
- namespace.supervisorsvc.version.rm

Don't have a strong opinion here, mostly thinking loud! Maybe namespace.service? (

embano1 replied:
Yeah, `namespace.service` or even `namespace.svc`.

rikatz replied:
Changed the command here to svc. 

I left the package name as supervisorsvc as this is the internal representation in the API, but the whole command now will be namespace.svc.

embano1 commented on vapi/namespace/supervisorsvc.go:
```diff
@@ -0,0 +1,101 @@
+
+// ActivateSupervisorServices activates a previously registered Supervisor Service.
+func (c *Manager) ActivateSupervisorServices(ctx context.Context, id string) error {
+	var response interface{}
```
question: why `interface{}`?

rikatz replied:
copy/paste :) tbh I guess this can be a string as the response will always be empty anyway

rikatz replied:
moved to nil response

dougm commented on govc/namespace/service/create.go:
```diff
@@ -0,0 +1,94 @@
+func (cmd *createService) Run(ctx context.Context, f *flag.FlagSet) error {
+	manifest := f.Args()
+	if len(manifest) != 1 {
+		return fmt.Errorf("this command requires exactly one argument")
```
Add:
```go
func (cmd *create) Usage() string {
    return "MANIFEST"
}
```
And changing this to return `flag.ErrHelp` should be clear usage.

dougm commented on vapi/simulator/simulator.go:
```diff
@@ -291,6 +291,21 @@ func (s *handler) DetachTag(id vim.ManagedObjectReference, tag vim.VslmTagEntry)
 
+// StatusCustom responds with custom HTTP status and json encoded val if given.
+// For use with "/api" endpoints.
+func StatusCustom(w http.ResponseWriter, httpStatus int, val ...interface{}) {
```
Could just name this `Status` and change `StatusOK` to use it:
```go
func StatusOK(w http.ResponseWriter, val ...interface{}) {
	Status(w, http.StatusOK, val...)
}
```

rikatz replied:
yeah good catch! didn't wanted to mess with already working things :D changed here

dougm reviewed (commented):
Thank you @rikatz , looks great! Just a few minor suggestions/requests.

rikatz commented on govc/namespace/service/activate.go:
```diff
@@ -0,0 +1,68 @@
+	return `Activates a vSphere Namespace Supervisor Service.
+
+Examples:
+  govc namespace.supervisorsvc.activate my-supervisor-service other-supervisor-service`
```
I have renamed all to "service" then

rikatz commented on govc/namespace/service/activate.go:
```diff
@@ -0,0 +1,68 @@
+	"github.com/vmware/govmomi/vapi/namespace"
+)
+
+type activateService struct {
```
ok, did in all commands with leading Service in struct

rikatz commented on govc/USAGE.md:
```diff
@@ -230,6 +230,12 @@ but appear via `govc $cmd -h`:
  - [namespace.cluster.enable](#namespaceclusterenable)
  - [namespace.cluster.ls](#namespaceclusterls)
  - [namespace.logs.download](#namespacelogsdownload)
+ - [namespace.svc.activate](#namespacesvcactivate)
```
done :)

rikatz commented on govc/namespace/service/activate.go:
```diff
@@ -0,0 +1,68 @@
+func (cmd *activateService) Run(ctx context.Context, f *flag.FlagSet) error {
+	services := f.Args()
+	if len(services) < 1 {
+		return fmt.Errorf("at least one service must be passed as argument")
```
Right, thanks! I'm not sure I should write something else in NAME... or just leave the NAME... as is. So I left as is (exactly as in the patch!)

rikatz commented:
@dougm thanks for the review :) and sorry for the delay, a lot of other stuff ongoing here :/

Hopefully this is right now and all the tests will pass as well :)

embano1 commented:
Thx @rikatz ! Once the review is done, can you please squash your commits and follow our contribution guidelines in terms of title and message body (reference issue) so your work renders nicely in the CHANGELOG?

rikatz commented:
@embano1 happy holidays :) sorry for the delay, last days were busy!

I have squashed the commits and rewritten the commit message to be aligned with contribution guideline.

Thanks!

embano1 commented:
Thx, LGTM but will ask @dougm to approve.

dougm reviewed (approved):
Looks great @rikatz , thanks!
